fuzz/fuzz_targets/tcp_headers.rs

```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+impl smoltcp::phy::Fuzzer for TcpHeaderFuzzer {
+    fn fuzz_packet(&self, frame_data: &mut [u8]) {
+        if self.1 == 0 {
+            return;
+        }
+
+        let tcp_packet_offset = {
+            let mut eth_frame = EthernetFrame::new_unchecked(&frame_data);
+            EthernetFrame::<&mut [u8]>::header_len() + match eth_frame.ethertype() {
+                EthernetProtocol::Ipv4 =>
+                    Ipv4Packet::new_unchecked(eth_frame.payload()).header_len() as usize,
+                EthernetProtocol::Ipv6 =>
+                    Ipv6Packet::new_unchecked(eth_frame.payload()).header_len() as usize,
+                _ => return
+            }
+        };
+
+        let tcp_is_syn = {
+            let tcp_packet = TcpPacket::new_checked(&frame_data[tcp_packet_offset..]).unwrap();
+            tcp_packet.syn()
+        };
+
+        if tcp_is_syn {
+            return;
+        }
+
+        if !frame_data.ends_with(b"abcdef") {
+            return;
+        }
+
+        let tcp_header_len = {
+            let tcp_packet = &frame_data[tcp_packet_offset..];
+            (tcp_packet[12] as usize >> 4) * 4
+        };
+
+        let tcp_packet = &mut frame_data[tcp_packet_offset+4..];
+
+        let replacement_data = &self.0[..self.1];
+        let copy_len = cmp::min(replacement_data.len(), tcp_header_len);
+        assert!(copy_len < tcp_packet.len());
+        tcp_packet[..copy_len].copy_from_slice(&replacement_data[..copy_len]);
+    }
+}
+
+struct EmptyFuzzer();
+
+impl smoltcp::phy::Fuzzer for EmptyFuzzer {
+    fn fuzz_packet(&self, _: &mut [u8]) {}
+}
+
+fuzz_target!(|data: &[u8]| {
+    let clock = mock::Clock::new();
+
+    let device = {
+
+        let (mut opts, mut free) = utils::create_options();
+        utils::add_middleware_options(&mut opts, &mut free);
+
+        let mut matches = utils::parse_options(&opts, free);
+        let device = utils::parse_middleware_options(&mut matches, Loopback::new(),
+                                                     /*loopback=*/true);
+
+        smoltcp::phy::FuzzInjector::new(device,
+                                        EmptyFuzzer(),
+                                        TcpHeaderFuzzer::new(data))
+    };
+
+    let mut neighbor_cache_entries = [None; 8];
+    let neighbor_cache = NeighborCache::new(&mut neighbor_cache_entries[..]);
+
+    let ip_addrs = [IpCidr::new(IpAddress::v4(127, 0, 0, 1), 8)];
+    let mut iface = EthernetInterfaceBuilder::new(device)
+            .ethernet_addr(EthernetAddress::default())
+            .neighbor_cache(neighbor_cache)
+            .ip_addrs(ip_addrs)
+            .finalize();
+
+    let server_socket = {
+        // It is not strictly necessary to use a `static mut` and unsafe code here, but
+        // on embedded systems that smoltcp targets it is far better to allocate the data
+        // statically to verify that it fits into RAM rather than get undefined behavior
+        // when stack overflows.
```

Comment:
I'm confused. The fuzzer is not an embedded target, is it?

Reply:
This was copy-pasted from the loopback example.